### PR TITLE
object_isinst

### PR DESCRIPTION
**Pull request by mrvoorhe**
Not complete yet.  Opening to ask some questions.

### CODE REVIEW

File: unity/unity-embed-host/CoreCLRHost.cs [hunk 1]
```diff
@@ -95,6 +95,34 @@ public static IntPtr gchandle_get_target_v2([NativeWrapperType("uintptr_t")] Int
...
     }
 
+    [return: NativeCallbackType("MonoObject*")]
+    public static IntPtr object_isinst([NativeCallbackType("MonoObject*")] IntPtr obj, [NativeCallbackType("MonoClass*")] IntPtr klass)
+    {
+        var instance = Unsafe.As<IntPtr, object>(ref obj);
+        var type = Unsafe.As<IntPtr, Type>(ref klass);
+        Type current = instance.GetType();
+
+        while (current != null)
```
Comment (mrvoorhe) [inline unity/unity-embed-host/CoreCLRHost.cs:105]: @joncham @UnityAlex I recall we discussed for some methods there will be a slow way and maybe faster ways.   Does this reflection based implementation fall under the "slow way category"?  Is there a different approach to take that is faster?

File: unity/unity-embed-host/CoreCLRHost.cs [hunk 2]
```diff
@@ -95,6 +95,34 @@ public static IntPtr gchandle_get_target_v2([NativeWrapperType("uintptr_t")] Int
...
+            foreach (var iface in current.GetInterfaces())
+            {
+                if (iface == type)
+                    return obj;
+            }
+
+            current = current.BaseType;
+        }
+
+        return nint.Zero;
```
Comment (mrvoorhe) [inline unity/unity-embed-host/CoreCLRHost.cs:123]: The old native implementation returned `NULL` when the object was not of the given type.  Does does anyone see a way to get `mono_object_isinst` to return `NULL` from here?

As this code is now, I have a test failure.  I'm trying to figure out the cleanest way to address this.
```
D:\UnitySrc\dev\unity-runtime\unity\embed_api_tests\main.cpp(513): FAILED:
  CHECK( mono_object_isinst(testobj, klass) != 0 )
with expansion:
  NULL != 0
```

File: unity/unity-embed-host/CoreCLRHost.cs [hunk 3]
```diff
@@ -95,6 +95,34 @@ public static IntPtr gchandle_get_target_v2([NativeWrapperType("uintptr_t")] Int
...
+        var instance = Unsafe.As<IntPtr, object>(ref obj);
+        var type = Unsafe.As<IntPtr, Type>(ref klass);
+        Type current = instance.GetType();
+
+        while (current != null)
+        {
+            if (current == type)
+                return obj;
+
+            // TODO : Unclear what the value of this is.  The if above is always hit first preventing rank differences from being taken into consideration
```
Comment (mrvoorhe) [inline unity/unity-embed-host/CoreCLRHost.cs:110]: I haven't figured out what needs to happen here yet.  Based on reading the native implmentation https://github.com/Unity-Technologies/runtime/blob/20cf508e82ba7cce3ccb141acdaef2a3d0c7d4ad/src/coreclr/vm/mono/mono_coreclr.cpp#L969

I thought that test cases such as `[TestCase(typeof(Mammal), 0, typeof(Mammal), 1, false)]` should fail due to the rank being different.  However,

1) In the managed implementation the `if (current == type)` above is hit during this case and we say they are equal
2) I don't see any tests in `main.cpp` that test the array behavior of the current native implementation.

File: unity/unity-embed-host/CoreCLRHost.cs [hunk 4]
```diff
@@ -95,6 +95,34 @@ public static IntPtr gchandle_get_target_v2([NativeWrapperType("uintptr_t")] Int
         return Unsafe.As<object, IntPtr>(ref obj);
     }
 
+    [return: NativeCallbackType("MonoObject*")]
+    public static IntPtr object_isinst([NativeCallbackType("MonoObject*")] IntPtr obj, [NativeCallbackType("MonoClass*")] IntPtr klass)
+    {
+        var instance = Unsafe.As<IntPtr, object>(ref obj);
+        var type = Unsafe.As<IntPtr, Type>(ref klass);
```
Comment (joncham) [inline unity/unity-embed-host/CoreCLRHost.cs:102]: `klass` will be a pointer to native metadata not a `System.Type` instance. It should be compatible with RuntimeTypeHandle if things are setup correctly.

So you want something like:
```
Type.GetTypeFromHandle(RuntimeTypeHandle.FromIntPtr(klass))
```

File: unity/unity-embed-host/CoreCLRHost.cs (hunk 1 above)
Comment (joncham) [reply unity/unity-embed-host/CoreCLRHost.cs:105]: You may be able to call `Type.IsAssignableFrom` https://learn.microsoft.com/en-us/dotnet/api/system.type.isassignablefrom?view=net-7.0

File: unity/unity-embed-host/CoreCLRHost.cs (hunk 2 above)
Comment (joncham) [reply unity/unity-embed-host/CoreCLRHost.cs:123]: I think your error is that you are returning 0/NULL when you should not. That test is asserting for not 0.

File: unity/unity-embed-host/CoreCLRHost.cs (hunk 1 above)
Comment (UnityAlex) [reply unity/unity-embed-host/CoreCLRHost.cs:105]: I believe the performance concern was more focused on object creation. The theory was that doing object creation through reflection would be slower than trying to stick to native or some other approach.

**Comment by mrvoorhe** [comment]
I'm not sure what the desired behavior for some array cases is.
```
      IsInstArrays(UnityEmbedHost.Tests.Mammal,1,UnityEmbedHost.Tests.Anaimal,1,False)
      IsInstArrays(UnityEmbedHost.Tests.Mammal,1,UnityEmbedHost.Tests.IAnimal,1,False)
      IsInstArrays(UnityEmbedHost.Tests.Mammal,1,UnityEmbedHost.Tests.IMammal,1,False)
```

These tests currently fail when I implement `isinst` using `IsAssignableTo`. 

I came to the conclusion `isinst` should return false (i.e. NULL) based on my understanding of the current native method  `mono_class_is_subclass_of`.  I could be wrong about that, there are no native tests using arrays.

This scenario essentially boils down to co-variance.  Ex:
```
        var m = new Mammal[] { new Mammal(), new Mammal() };
        IMammal[] i = m;
```

`IsAssignableTo` is going to return true, because you can assign `m` to i`.  The question is, should `isinst` report the same as `IsAssignableTo`?  Or is `isinst` aiming for a slightly different behavior?

fyi - I don't have these changes pushed.  I can clean up and push the changes if anyone would like to see the code.

@joncham @UnityAlex What are your thoughts?

**Comment by joncham** [comment]
> What are your thoughts?

We can check mono embedding API, but my feeling is this should return true (e.g. it should be true for the arrays). I would naively expect the isinst calls to succeed for any assignment/cast in C# that will succeed.

**Comment by joncham** [comment]
I think mono even falls back to the isinst logic for the slow/complex cases of casting.

**Comment by mrvoorhe** [comment]
I'm going to close this PR.  I tired to keep `objcet_isinst` and `object_get_class` separate but I ended up mixing changes while fixing a few issues and doing clean up.  I don't think it's worth the effort to pull the changes back apart again.

**Pull request closed without merging**